tests/test_fetch.py

```diff
@@ -0,0 +1,402 @@
+        for c, l in list(zip(cur, languages))[:4]:
+            assert np.all(
+                [cc == ll for cc, ll in zip(c, l)]
+            ), "Sorting order is different"
+
+    def test_head_tail(self, schema_any):
+        query = schema.User * schema.Language
+        n = 5
+        frame = query.head(n, format="frame")
+        assert isinstance(frame, pandas.DataFrame)
+        array = query.head(n, format="array")
+        assert array.size == n
+        assert len(frame) == n
+        assert query.primary_key == frame.index.names
+
+        n = 4
+        frame = query.tail(n, format="frame")
+        array = query.tail(n, format="array")
+        assert array.size == n
+        assert len(frame) == n
+        assert query.primary_key == frame.index.names
+
+    def test_limit_offset(self, schema_any, lang, languages):
+        """Test the limit and offset kwargs together"""
+        cur = lang.fetch(offset=2, limit=4, order_by=["language", "name DESC"])
+        languages.sort(key=itemgetter(0), reverse=True)
+        languages.sort(key=itemgetter(1), reverse=False)
+        assert len(cur) == 4, "Length is not correct"
+        for c, l in list(zip(cur, languages[2:6])):
+            assert np.all(
+                [cc == ll for cc, ll in zip(c, l)]
+            ), "Sorting order is different"
+
+    def test_iter(self, schema_any, lang, languages):
+        """Test iterator"""
+        cur = lang.fetch(order_by=["language", "name DESC"])
+        languages.sort(key=itemgetter(0), reverse=True)
+        languages.sort(key=itemgetter(1), reverse=False)
+        for (name, lang_val), (tname, tlang) in list(zip(cur, languages)):
+            assert name == tname and lang_val == tlang, "Values are not the same"
+        # now as dict
+        cur = lang.fetch(as_dict=True, order_by=("language", "name DESC"))
+        for row, (tname, tlang) in list(zip(cur, languages)):
+            assert (
+                row["name"] == tname and row["language"] == tlang
+            ), "Values are not the same"
+
+    def test_keys(self, schema_any, lang, languages):
+        """test key fetch"""
+        languages.sort(key=itemgetter(0), reverse=True)
+        languages.sort(key=itemgetter(1), reverse=False)
+
+        lang = schema.Language()
+        cur = lang.fetch("name", "language", order_by=("language", "name DESC"))
+        cur2 = list(lang.fetch("KEY", order_by=["language", "name DESC"]))
+
+        for c, c2 in zip(zip(*cur), cur2):
+            assert c == tuple(c2.values()), "Values are not the same"
+
+    def test_attributes_as_dict(self, schema_any, subject):
+        """
+        Issue #595
+        """
+        attrs = ("species", "date_of_birth")
+        result = subject.fetch(*attrs, as_dict=True)
+        assert bool(result) and len(result) == len(subject)
+        assert set(result[0]) == set(attrs)
+
+    def test_fetch1_step1(self, schema_any, lang, languages):
+        assert (
```

Comment:
Added this assertion to check if upstream tests like `test_offset_by` changed the `contents` and broke this test.